casper/src/main/scala/coop/rchain/casper/util/TopologicalSortUtil.scala

```diff
@@ -0,0 +1,26 @@
+package coop.rchain.casper.util
+
+import coop.rchain.casper.Estimator.BlockHash
+import coop.rchain.casper.protocol.BlockMessage
+import coop.rchain.casper.util.ProtoUtil.blockNumber
+
+object TopologicalSortUtil {
+  type BlockSort = Vector[Vector[BlockHash]]
+  def update(sort: BlockSort, offset: Long, block: BlockMessage): BlockSort = {
+    val hash   = block.blockHash
+    val number = (blockNumber(block) - offset).toInt
+
+    if (number == sort.length) {
+      //this is a new block height
+      sort :+ Vector(hash)
+    } else if (number < sort.length) {
+      //this is another block at a known height
+      val curr = sort(number)
+      sort.updated(number, curr :+ hash)
+    } else {
+      //impossible because block numbers must be sequential
+      throw new Exception("Attempted to add block with invalid block number to state.")
```

Comment:
My understanding is that this exception can only happen in case of an algorithm error. 

If this is correct then we should assert  that `number <= sort.length` as a function precondition such that we make it clear that it's due to a bug in the algorithm.